[BUG] `Unable to expand resources because the source file ... The function "references" was not found.`  when using languageVersion 2.0

### Existing rule

_No response_

### Description of the issue

When referencing the outputs of a collection of deployed modules, it seems that PSRule is unable to recognize the `references` function used to access the properties of the modules.

### Error messages

##[error]Failed to expand bicep source '/home/vsts/work/1/s/main.bicep'. Exception calling "GetBicepResources" with "2" argument(s): "Unable to expand resources because the source file '/home/vsts/work/1/s/main.bicep' was not valid. An error occurred evaluating expression '[map(references('acc_loop'), lambda('acc', createObject(format('{0}', lambdaVariables('acc').outputs.acc.value), createObject('name', lambdaVariables('acc').outputs.acc.value))))]' line 257. The function "references" was not found."


### Reproduction

module1.bicep

```bicep
param acc int = 0

metadata name = 'Create a Windows Virtual Machine'

//
// Parameters
//
@minLength(1)
@maxLength(64)
@description('Name of the Virtual Machine')
param vm_name string = replace(resourceGroup().name, '-rg-', '-vm-')

@description('Location of the Virtual Machine')
param location string = resourceGroup().location

@description('Tags for the Virtual Machine')
param tags object = resourceGroup().tags

@description('Id of the Network Interface to attach to the Virtual Machine')
param network_interface_id string

@minLength(1)
@maxLength(16)
@description('Username for Virtual Machine admin')
param admin_username string

@description('Admin password')
@secure()
param admin_password string

@minLength(1)
@maxLength(80)
@description('Name of the Operating System Disk for the Virtual Machine')
param os_disk_name string = replace(resourceGroup().name, '-rg-', '-osdisk-')

@description('ID of the Data Disk for the Virtual Machine')
param data_disk_id string

@minLength(1)
@maxLength(80)
@description('Name of the Data Disk for the Virtual Machine')
param data_disk_name string

@description('Virtual Machine Size')
param vm_size string = 'Standard_D2s_v3'

@description('Size in GB of Operating System Disk')
param disk_size_gb int = 127

@description('Storage Account Type of Managed Disk')
@allowed([
  'Standard_LRS'
  'Premium_LRS'
  'Premium_ZRS'
  'StandardSSD_LRS'
  'StandardSSD_ZRS'
  'UltraSSD_LRS'
])
param storage_account_type string = 'StandardSSD_LRS'

@allowed([
  'None'
  'SystemAssigned'
])
param identity string = 'None'

//
// Resources
//
resource virtual_machine 'Microsoft.Compute/virtualMachines@2021-07-01' = {
  name: vm_name
  location: location
  tags: tags
  identity: {
    type: identity
  }
  properties: {
    hardwareProfile: {
      vmSize: vm_size
    }

    networkProfile: {
      networkInterfaces: [
        {
          id: network_interface_id
           properties: {
             deleteOption: 'Detach'
           }
        }
      ]
    }

    osProfile: {
      adminUsername: admin_username
      adminPassword: admin_password
      computerName: take(vm_name, 15)
      windowsConfiguration: {
        enableAutomaticUpdates: true
        provisionVMAgent: true
        timeZone: 'Romance Standard Time'
      }
    }

    storageProfile: {
      osDisk: {
        createOption: 'FromImage'
        deleteOption: 'Detach'
        diskSizeGB: disk_size_gb
        managedDisk: {
          storageAccountType: storage_account_type
        }
        name: os_disk_name
        osType: 'Windows'
      }
      dataDisks: [
        {
          lun: 0
          createOption: 'Attach'
          deleteOption: 'Detach'
          caching: 'ReadOnly'
          writeAcceleratorEnabled: false
          managedDisk: {
            id: data_disk_id
          }
          diskSizeGB: 40
          name: data_disk_name
        }
      ]

      imageReference: {
        publisher: 'MicrosoftWindowsServer'
        offer: 'WindowsServer'
        sku: '2022-datacenter-azure-edition'
        version: 'latest'
      }
    }
    licenseType: 'Windows_Server'
  }
}

@description('Name of the created VM')
output output_vm_name string = virtual_machine.name


output acc string = guid(string(acc))
```

main.bicep

```bicep
module acc_loop 'module1.bicep' = [for (item, index) in range(0, 5): {
  name: 'deploy-${index}'
  params: {
    acc: index
    admin_username: 'sqdsqdqqs'
    admin_password: uniqueString('sqdsqdsqsq')
    data_disk_id: 'sqdsqsq'
    data_disk_name: 'sqdsqdssq'
    network_interface_id: 'qsdsdq'
  }

}]

var output_array = map(acc_loop, acc => {
  '${acc.outputs.acc}': {
    name: acc.outputs.acc
  }
})

output output array = output_array
```

The compoilation of main.bicep fails with:

```
##[error]Failed to expand bicep source '/home/vsts/work/1/s/main.bicep'. Exception calling "GetBicepResources" with "2" argument(s): "Unable to expand resources because the source file '/home/vsts/work/1/s/main.bicep' was not valid. An error occurred evaluating expression '[map(references('acc_loop'), lambda('acc', createObject(format('{0}', lambdaVariables('acc').outputs.acc.value), createObject('name', lambdaVariables('acc').outputs.acc.value))))]' line 257. The function "references" was not found."
```

### Version of PSRule

2.9.0

### Version of PSRule for Azure

1.37.0

### Additional context

_No response_

--- a/bicepconfig.json
+++ b/bicepconfig.json
@@ -1,6 +1,5 @@
 {
   "experimentalFeaturesEnabled": {
-    "optionalModuleNames": true,
-    "userDefinedFunctions": true
+    "optionalModuleNames": true
   }
 }

--- a/docs/CHANGELOG-v1.md
+++ b/docs/CHANGELOG-v1.md
@@ -34,6 +34,9 @@ What's changed since pre-release v1.38.0-B0034:
 - Engineering:
   - Quality updates to rule documentation by @BernieWhite.
     [#2570](https://github.com/Azure/PSRule.Rules.Azure/issues/2570)
+- Bug fixes:
+  - Fixed support for `references` function by @BernieWhite.
+    [#2922](https://github.com/Azure/PSRule.Rules.Azure/issues/2922)
 
 ## v1.38.0-B0034 (pre-release)
 

--- a/src/PSRule.Rules.Azure/Data/Policy/PolicyAssignmentVisitor.cs
+++ b/src/PSRule.Rules.Azure/Data/Policy/PolicyAssignmentVisitor.cs
@@ -220,6 +220,12 @@ namespace PSRule.Rules.Azure.Data.Policy
                 return false;
             }
 
+            public bool TryGetResourceCollection(string symbolicName, out IResourceValue[] resources)
+            {
+                resources = null;
+                return false;
+            }
+
             internal bool TryParameterAssignment(string parameterName, out JToken value)
             {
                 return _ParameterAssignments.TryGetValue(parameterName, out value);

--- a/src/PSRule.Rules.Azure/Data/Template/ArrayDeploymentSymbol.cs
+++ b/src/PSRule.Rules.Azure/Data/Template/ArrayDeploymentSymbol.cs
@@ -1,13 +1,14 @@
-﻿// Copyright (c) Microsoft Corporation.
+// Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+using System;
 using System.Collections.Generic;
 
 namespace PSRule.Rules.Azure.Data.Template
 {
     internal sealed class ArrayDeploymentSymbol : DeploymentSymbol, IDeploymentSymbol
     {
-        private List<ResourceValue> _Resources;
+        private List<string> _Ids;
 
         public ArrayDeploymentSymbol(string name)
             : base(name) { }
@@ -16,13 +17,18 @@ namespace PSRule.Rules.Azure.Data.Template
 
         public void Configure(ResourceValue resource)
         {
-            _Resources ??= new List<ResourceValue>();
-            _Resources.Add(resource);
+            _Ids ??= new List<string>();
+            _Ids.Add(resource.Id);
         }
 
         public string GetId(int index)
         {
-            return _Resources[index].Id;
+            return _Ids[index];
+        }
+
+        public string[] GetIds()
+        {
+            return _Ids?.ToArray() ?? Array.Empty<string>();
         }
     }
 }

--- a/src/PSRule.Rules.Azure/Data/Template/Functions.cs
+++ b/src/PSRule.Rules.Azure/Data/Template/Functions.cs
@@ -113,6 +113,7 @@ namespace PSRule.Rules.Azure.Data.Template
             new FunctionDescriptor("pickZones", PickZones),
             new FunctionDescriptor("providers", Providers),
             new FunctionDescriptor("reference", Reference),
+            new FunctionDescriptor("references", References),
             new FunctionDescriptor("resourceId", ResourceId),
             new FunctionDescriptor("subscriptionResourceId", SubscriptionResourceId),
             new FunctionDescriptor("tenantResourceId", TenantResourceId),
@@ -1011,6 +1012,37 @@ namespace PSRule.Rules.Azure.Data.Template
             return context.TryGetResource(resourceId, out var resourceValue)
                 ? GetReferenceResult(resourceValue, full)
                 : full ? new Mock.MockResource(resourceId) : new Mock.MockResource(resourceId)["properties"];
+        }
+
+        /// <summary>
+        /// references(symbolic name of a resource collection, ['Full', 'Properties'])
+        /// </summary>
+        /// <remarks>
+        /// See <seealso href="https://learn.microsoft.com/azure/azure-resource-manager/templates/template-functions-resource#references"/>.
+        /// </remarks>
+        internal static object References(ITemplateContext context, object[] args)
+        {
+            var argCount = CountArgs(args);
+            if (argCount is < 1 or > 2)
+                throw ArgumentsOutOfRange(nameof(References), args);
+
+            string fullValue = null;
+            var full = argCount == 2 && ExpressionHelpers.TryString(args[1], out fullValue) && string.Equals(fullValue, PROPERTY_FULL, StringComparison.OrdinalIgnoreCase);
+            if (argCount == 2 && !full && !string.Equals(fullValue, PROPERTY_PROPERTIES, StringComparison.OrdinalIgnoreCase))
+                throw ArgumentFormatInvalid(nameof(References));
+
+            // Get symbolic name
+            if (!ExpressionHelpers.TryString(args[0], out var symbolicName))
+                throw ArgumentFormatInvalid(nameof(References));
+
+            if (!context.TryGetResourceCollection(symbolicName, out var resources))
+                throw ArgumentInvalidResourceCollection(nameof(References), symbolicName);
+
+            var result = new object[resources.Length];
+            for (var i = 0; i < resources.Length; i++)
+                result[i] = GetReferenceResult(resources[i], full);
+
+            return result;
         }
 
         private static object GetReferenceResult(IResourceValue resource, bool full)
@@ -2479,6 +2511,11 @@ namespace PSRule.Rules.Azure.Data.Template
                 expression,
                 string.Format(Thread.CurrentThread.CurrentCulture, PSRuleResources.ArgumentNullNotExpected, expression)
             );
+        }
+
+        private static Exception ArgumentInvalidResourceCollection(string v, string symbolicName)
+        {
+            throw new NotImplementedException();
         }
 
         #endregion Exceptions

--- a/src/PSRule.Rules.Azure/Data/Template/ITemplateContext.cs
+++ b/src/PSRule.Rules.Azure/Data/Template/ITemplateContext.cs
@@ -57,8 +57,27 @@ namespace PSRule.Rules.Azure.Data.Template
 
         bool TryDefinition(string type, out ITypeDefinition definition);
 
-        bool TryGetResource(string resourceId, out IResourceValue resource);
+        /// <summary>
+        /// Try to get a resource from the current context.
+        /// </summary>
+        /// <param name="nameOrResourceId">The symbolic name or resource identifier.</param>
+        /// <param name="resource">A resource.</param>
+        /// <returns>Returns <c>true</c> when the resource exists.</returns>
+        bool TryGetResource(string nameOrResourceId, out IResourceValue resource);
 
+        /// <summary>
+        /// Try to get a resource collection from the current context.
+        /// </summary>
+        /// <param name="symbolicName">The symbolic name for the collection.</param>
+        /// <param name="resources">A collection of resources.</param>
+        /// <returns>Returns <c>true</c> when the symbolic name exists.</returns>
+        bool TryGetResourceCollection(string symbolicName, out IResourceValue[] resources);
+
+        /// <summary>
+        /// Write a debug message for the current execution context.
+        /// </summary>
+        /// <param name="message">The format message.</param>
+        /// <param name="args">Additional arguments for the format message.</param>
         void WriteDebug(string message, params object[] args);
 
         /// <summary>

--- a/src/PSRule.Rules.Azure/Data/Template/NestedTemplateContext.cs
+++ b/src/PSRule.Rules.Azure/Data/Template/NestedTemplateContext.cs
@@ -57,11 +57,19 @@ namespace PSRule.Rules.Azure.Data.Template
             return _Inner.TryVariable(variableName, out value);
         }
 
-        public bool TryGetResource(string resourceId, out IResourceValue resource)
+        /// <inheritdoc/>
+        public bool TryGetResource(string nameOrResourceId, out IResourceValue resource)
         {
-            return _Inner.TryGetResource(resourceId, out resource);
+            return _Inner.TryGetResource(nameOrResourceId, out resource);
         }
 
+        /// <inheritdoc/>
+        public bool TryGetResourceCollection(string symbolicName, out IResourceValue[] resources)
+        {
+            return _Inner.TryGetResourceCollection(symbolicName, out resources);
+        }
+
+        /// <inheritdoc/>
         public void WriteDebug(string message, params object[] args)
         {
             _Inner.WriteDebug(message, args);

--- a/src/PSRule.Rules.Azure/Data/Template/ObjectDeploymentSymbol.cs
+++ b/src/PSRule.Rules.Azure/Data/Template/ObjectDeploymentSymbol.cs
@@ -1,11 +1,11 @@
-﻿// Copyright (c) Microsoft Corporation.
+// Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
 namespace PSRule.Rules.Azure.Data.Template
 {
     internal sealed class ObjectDeploymentSymbol : DeploymentSymbol, IDeploymentSymbol
     {
-        private ResourceValue _Resource;
+        private string _ResourceId;
 
         public ObjectDeploymentSymbol(string name, ResourceValue resource)
             : base(name)
@@ -18,12 +18,12 @@ namespace PSRule.Rules.Azure.Data.Template
 
         public void Configure(ResourceValue resource)
         {
-            _Resource = resource;
+            _ResourceId = resource.Id;
         }
 
         public string GetId(int index)
         {
-            return _Resource.Id;
+            return _ResourceId;
         }
     }
 }

--- a/src/PSRule.Rules.Azure/Data/Template/TemplateVisitor.cs
+++ b/src/PSRule.Rules.Azure/Data/Template/TemplateVisitor.cs
@@ -221,16 +221,32 @@ namespace PSRule.Rules.Azure.Data.Template
                 _ResourceIds.Remove(resource.Id);
             }
 
-            public bool TryGetResource(string resourceId, out IResourceValue resource)
+            /// <inheritdoc/>
+            public bool TryGetResource(string nameOrResourceId, out IResourceValue resource)
             {
-                if (_Symbols.TryGetValue(resourceId, out var symbol))
-                    resourceId = symbol.GetId(0);
+                if (_Symbols.TryGetValue(nameOrResourceId, out var symbol))
+                    nameOrResourceId = symbol.GetId(0);
 
-                if (_ResourceIds.TryGetValue(resourceId, out resource))
+                if (_ResourceIds.TryGetValue(nameOrResourceId, out resource))
                     return true;
 
                 resource = null;
                 return false;
+            }
+
+            /// <inheritdoc/>
+            public bool TryGetResourceCollection(string symbolicName, out IResourceValue[] resources)
+            {
+                resources = null;
+                if (!_Symbols.TryGetValue(symbolicName, out var symbol) || symbol is not ArrayDeploymentSymbol array)
+                    return false;
+
+                var ids = array.GetIds();
+                resources = new IResourceValue[ids.Length];
+                for (var i = 0; i < ids.Length; i++)
+                    resources[i] = _ResourceIds[ids[i]];
+
+                return true;
             }
 
             public void AddOutput(string name, JObject output)

--- a/tests/PSRule.Rules.Azure.Tests/Bicep/SymbolicNameTestCases/Tests.Bicep.1.bicep
+++ b/tests/PSRule.Rules.Azure.Tests/Bicep/SymbolicNameTestCases/Tests.Bicep.1.bicep
@@ -1,0 +1,16 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// Test case for https://github.com/Azure/PSRule.Rules.Azure/issues/2922
+
+module child_loop './Tests.Bicep.1.child.bicep' = [
+  for (item, index) in range(0, 2): {
+    name: 'deploy-${index}'
+    params: {
+      index: index
+    }
+  }
+]
+
+output items array = map(child_loop, item => item.outputs.childValue)
+output itemsAsString string[] = map(child_loop, item => item.outputs.childValue)

--- a/tests/PSRule.Rules.Azure.Tests/Bicep/SymbolicNameTestCases/Tests.Bicep.1.child.bicep
+++ b/tests/PSRule.Rules.Azure.Tests/Bicep/SymbolicNameTestCases/Tests.Bicep.1.child.bicep
@@ -1,0 +1,6 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+param index int
+
+output childValue string = 'child-${index}'

--- a/tests/PSRule.Rules.Azure.Tests/Bicep/SymbolicNameTestCases/Tests.Bicep.1.json
+++ b/tests/PSRule.Rules.Azure.Tests/Bicep/SymbolicNameTestCases/Tests.Bicep.1.json
@@ -1,0 +1,71 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+  "languageVersion": "2.0",
+  "contentVersion": "1.0.0.0",
+  "metadata": {
+    "_generator": {
+      "name": "bicep",
+      "version": "0.28.1.47646",
+      "templateHash": "1381069533852459453"
+    }
+  },
+  "resources": {
+    "child_loop": {
+      "copy": {
+        "name": "child_loop",
+        "count": "[length(range(0, 2))]"
+      },
+      "type": "Microsoft.Resources/deployments",
+      "apiVersion": "2022-09-01",
+      "name": "[format('deploy-{0}', copyIndex())]",
+      "properties": {
+        "expressionEvaluationOptions": {
+          "scope": "inner"
+        },
+        "mode": "Incremental",
+        "parameters": {
+          "index": {
+            "value": "[copyIndex()]"
+          }
+        },
+        "template": {
+          "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
+          "languageVersion": "2.0",
+          "contentVersion": "1.0.0.0",
+          "metadata": {
+            "_generator": {
+              "name": "bicep",
+              "version": "0.28.1.47646",
+              "templateHash": "15231390219392886169"
+            }
+          },
+          "parameters": {
+            "index": {
+              "type": "int"
+            }
+          },
+          "resources": {},
+          "outputs": {
+            "childValue": {
+              "type": "string",
+              "value": "[format('child-{0}', parameters('index'))]"
+            }
+          }
+        }
+      }
+    }
+  },
+  "outputs": {
+    "items": {
+      "type": "array",
+      "value": "[map(references('child_loop'), lambda('item', lambdaVariables('item').outputs.childValue.value))]"
+    },
+    "itemsAsString": {
+      "type": "array",
+      "items": {
+        "type": "string"
+      },
+      "value": "[map(references('child_loop'), lambda('item', lambdaVariables('item').outputs.childValue.value))]"
+    }
+  }
+}

--- a/tests/PSRule.Rules.Azure.Tests/Bicep/SymbolicNameTestCases/bicepconfig.json
+++ b/tests/PSRule.Rules.Azure.Tests/Bicep/SymbolicNameTestCases/bicepconfig.json
@@ -1,0 +1,6 @@
+{
+  "experimentalFeaturesEnabled": {
+    "optionalModuleNames": true,
+    "symbolicNameCodegen": true
+  }
+}

--- a/tests/PSRule.Rules.Azure.Tests/FunctionTests.cs
+++ b/tests/PSRule.Rules.Azure.Tests/FunctionTests.cs
@@ -7,6 +7,7 @@ using System.Diagnostics;
 using System.Globalization;
 using System.IO;
 using System.Linq;
+using System.Security.Cryptography.Xml;
 using Newtonsoft.Json;
 using Newtonsoft.Json.Linq;
 using PSRule.Rules.Azure.Configuration;
@@ -808,6 +809,29 @@ namespace PSRule.Rules.Azure
             Assert.Equal(string.Empty, actual.Path);
             Assert.Equal(parentId, actual["id"].Value<string>());
             Assert.Equal("a", actual["name"].Value<string>());
+        }
+
+        [Fact]
+        [Trait(TRAIT, TRAIT_RESOURCE)]
+        public void References()
+        {
+            var context = GetContext();
+            var resourceId = "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/rg-test/providers/Unit.Test/type/a";
+            var resource = new ResourceValue(resourceId, "a-0", "Unit.Test/type", "child_loop[0]", new JObject(), null);
+            context.AddResource(resource);
+
+            var symbol = DeploymentSymbol.NewArray("child_loop");
+            symbol.Configure(resource);
+            context.AddSymbol(symbol);
+
+            var actual = (Functions.References(context, new object[] { "child_loop" }) as object[]).OfType<Mock.MockObject>().ToArray();
+            Assert.NotNull(actual);
+
+            actual = (Functions.References(context, new object[] { "child_loop", "Full" }) as object[]).OfType<Mock.MockObject>().ToArray();
+            Assert.NotNull(actual);
+
+            Assert.Throws<ExpressionArgumentException>(() => Functions.References(context, null));
+            Assert.Throws<ExpressionArgumentException>(() => Functions.References(context, new object[] { "Unit.Test/type", "test" }));
         }
 
         [Fact]

--- a/tests/PSRule.Rules.Azure.Tests/PSRule.Rules.Azure.Tests.csproj
+++ b/tests/PSRule.Rules.Azure.Tests/PSRule.Rules.Azure.Tests.csproj
@@ -260,6 +260,9 @@
     <None Update="Tests.Bicep.40.json">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
+    <None Update="Bicep\SymbolicNameTestCases\Tests.Bicep.1.json">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
     <None Update="Tests.Bicep.5.json">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>

--- a/tests/PSRule.Rules.Azure.Tests/TemplateVisitorTests.cs
+++ b/tests/PSRule.Rules.Azure.Tests/TemplateVisitorTests.cs
@@ -1270,7 +1270,7 @@ namespace PSRule.Rules.Azure
         [Fact]
         public void ProcessTemplate_WhenReferencesUsed_ReturnsItems()
         {
-            _ = ProcessTemplate(GetSourcePath("Bicep\\SymbolicNameTestCases\\Tests.Bicep.1.json"), null, out var templateContext);
+            _ = ProcessTemplate(GetSourcePath("Bicep/SymbolicNameTestCases/Tests.Bicep.1.json"), null, out var templateContext);
 
             Assert.True(templateContext.RootDeployment.TryOutput("items", out JObject output));
             var items = output["value"].Value<JArray>();

--- a/tests/PSRule.Rules.Azure.Tests/TemplateVisitorTests.cs
+++ b/tests/PSRule.Rules.Azure.Tests/TemplateVisitorTests.cs
@@ -1264,6 +1264,27 @@ namespace PSRule.Rules.Azure
             Assert.NotNull(items[1].Value<JObject>());
         }
 
+        /// <summary>
+        /// Test case for https://github.com/Azure/PSRule.Rules.Azure/issues/2922
+        /// </summary>
+        [Fact]
+        public void ProcessTemplate_WhenReferencesUsed_ReturnsItems()
+        {
+            _ = ProcessTemplate(GetSourcePath("Bicep\\SymbolicNameTestCases\\Tests.Bicep.1.json"), null, out var templateContext);
+
+            Assert.True(templateContext.RootDeployment.TryOutput("items", out JObject output));
+            var items = output["value"].Value<JArray>();
+
+            Assert.Equal("child-0", items[0].Value<string>());
+            Assert.Equal("child-1", items[1].Value<string>());
+
+            Assert.True(templateContext.RootDeployment.TryOutput("itemsAsString", out output));
+            items = output["value"].Value<JArray>();
+
+            Assert.Equal("child-0", items[0].Value<string>());
+            Assert.Equal("child-1", items[1].Value<string>());
+        }
+
         #region Helper methods
 
         private static string GetSourcePath(string fileName)

